### PR TITLE
Added support for authenticated proxy upstream

### PR DESCRIPTION
**Pull request by gbasto**

### CODE REVIEW

File: src/reqs.c [hunk 1]
```diff
@@ -1498,6 +1533,22 @@ void handle_connection (int fd)
 
         connptr->upstream_proxy = UPSTREAM_HOST (request->host);
         if (connptr->upstream_proxy != NULL) {
+                if (connptr->upstream_proxy->user)
+                {
+                        char proxy_auth[200] = "";
+                        char src[256];
+                        char dst2[512];
+                        strcpy(src, connptr->upstream_proxy->user);
```
Comment (rofl0r) [inline src/reqs.c]: this screams for the use of snprintf() over strcpy+strcat...

Comment (gbasto) [inline src/reqs.c]: Changed to snprintf(). Thank you for the feedback!

**Comment by rofl0r** [comment]
LGTM, though imo you should squash the 2 commits together.

File: src/reqs.c [hunk 2]
```diff
@@ -61,6 +61,41 @@
...
+        int n, l, i;
+        l = strlen(src);
+        max_len = (max_len - 1) / 4;
+        for (i = 0; i < max_len; i++, src += 3, l -= 3) {
+                switch (l) {
+                        case 0:
+                                break;
+                        case 1:
+                                n = src[0] << 16;
+                        *dest++ = base64[(n >> 18) & 077];
```
Comment (rofl0r) [inline src/reqs.c]: this is the base64 implementation taken from proxychains (of which i happen to be the maintainer), but it seems you made some white-space errors while copy/paste-ing it

File: src/reqs.c [hunk 3]
```diff
@@ -1498,6 +1533,19 @@ void handle_connection (int fd)
 
         connptr->upstream_proxy = UPSTREAM_HOST (request->host);
         if (connptr->upstream_proxy != NULL) {
+                if (connptr->upstream_proxy->user)
+                {
+                        char plain[256];
```
Comment (rofl0r) [inline src/reqs.c]: i think we could do with 2 buffers here, e.g. plain and proxy_auth could use the same buffer

**Pull request closed without merging**